### PR TITLE
Fix compliance issues with SysTime.to*String and SysTime.from*String.

### PR DESCRIPTION
**Pull request by jmdavis**
This fixes two issues:

* [issue# 15655](https://issues.dlang.org/show_bug.cgi?id=15655): H, H:MM, H:M, and HH:M (with either + or - in front) were incorrectly accepted as valid time zone strings by SysTime.from*String, when ISO 9601 requires HH or HH:MM (with either + or - in front). There also was no proper verification that the minutes were less than an hour.

* [issue# 15654](https://issues.dlang.org/show_bug.cgi?id=15654): SysTime.toISOString output the time zone in the extended format - HH:MM - rather than the non-extended format - HHMM. This fixes that.

Unfortunately, the fix for 15655 does risk breaking code for anyone accepting strings from elsewhere which didn't follow the ISO standard properly, but I don't know how to avoid that, and as long as the other sources for an ISO string do follow the standard, it's a non-issue. Nothing using only std.datetime to handle ISO strings will break and nothing interacting with correctly formatted ISO strings will break.

The fix for 15654 also risks breaking code, but in its case, we can mitigate the problem. `toISOString` now outputs the string in the correct format, but `fromISOString` accepts both the correct format and the old one with a warning in the docs that it will later be changed to only accept the correct format. So, any new strings that get written out will be fine, and the only risk should be if any strings outputted by `toISOString` prior to this change are kept around beyond the point that `fromISOString` is updated to only accept the correct format. And anyone who's outputted their strings in local time or UTC (which should be by far the most common) won't be affected at all.

### CODE REVIEW

**Comment by schveiguy** [comment]
Wow, this is a lot of code change. Hard to see where I should be focusing.

I'm wondering what it hurts to permanently accept the "incorrect" time zone in `fromISOString`. If people have written out timestamps into data files for a long period of time, it's not just a simple code change to fix that problem.

**Comment by jmdavis** [comment]
> Wow, this is a lot of code change. Hard to see where I should be focusing.

It might help to look at it commit by commit.

> I'm wondering what it hurts to permanently accept the "incorrect" time zone in fromISOString. If people have written out timestamps into data files for a long period of time, it's not just a simple code change to fix that problem.

How long to let `fromISOString` read the incorrectly formatted time zones is certainly debatable, because we do want to avoid causing anyone problems, but I believe very strongly that being lax with this sort of thing is just a source of problems in the long run. In general, if you're lax about what you accept, then you'll just keep getting junk, and no one will fix their code so that it actually does the right thing - they may not even realize that it's wrong.

So, while I'm certainly open to delaying fully fixing `fromISOString` for longer in order to minimize the risk of causing problems for folks who have been saving ISO strings that were not in local time or UTC, I definitely think that it would be a mistake not to fix it (though obviously, it would have been better had I never screwed this up in the first place).

Fortunately, in the vast majority of cases, folks are probably either going to have been using local time or UTC, and they're probably more likely to have been using the ISO extended format (since it's more legible, and IIRC, the RFC that relates to time strings matches the extended format, not the non-extended format). So, I would guess that in the vast majority of cases, folks won't have non-extended ISO strings with a numeric time zone written out anywhere to read in later, even if they are writing out non-extended ISO strings somewhere.

But I definitely need to get these changes in before I can fix it so that `LocalTime` includes the offset with the to*String functions. Otherwise, one of the most common time zones will suddenly be outputting incorrect strings from` toISOString`.

**Comment by schveiguy** [comment]
> if you're lax about what you accept, then you'll just keep getting junk, and no one will fix their code so that it actually does the right thing

It's not code we are talking about here. If I have GB of data that has this invalid timestamp string in it, then I have no recourse on reading that. I have to manually read that string, alter it to be "valid" and then pipe to `fromISOString`

I'd leave it as an undocumented "feature".

**Comment by jmdavis** [comment]
Sure, but then you don't catch bad data. Certainly, if it weren't for the faux pas with how the time zone was formatted in the first place, I don't think that there's any question that being lax about the format would be a bad idea.

Regardless, it's going to accept the incorrect format for now, and the documentation will point out the mistake so that folks have a chance to fix it. It's just a question of if/when we remove the workaround later.

**Comment by schveiguy** [comment]
> It might help to look at it commit by commit.

It doesn't. The code commits are buried in the commits to the unittests.

We really need to split up std.datetime...

File: std/datetime.d [hunk 1]
```diff
@@ -8013,10 +8013,20 @@ public:
...
         If this $(LREF SysTime)'s time zone is $(LREF LocalTime), then TZ is empty.
         If its time zone is $(D UTC), then it is "Z". Otherwise, it is the
-        offset from UTC (e.g. +1:00 or -7:00). Note that the offset from UTC
+        offset from UTC (e.g. +0100 or -0700). Note that the offset from UTC
         is $(I not) enough to uniquely identify the time zone.
 
-        Time zone offsets will be in the form +HH:MM or -HH:MM.
+        Time zone offsets will be in the form +HHMM or -HHMM.
+
+        Warning:
```
Comment (ZombineDev) [inline std/datetime.d]: Perhaps the `$(RED)` ddoc macro would be appropriate here.

**Comment by jmdavis** [comment]
> We really need to split up std.datetime...

Yes, which I need to find time to do in one chunk, since it doesn't work to do it over time. At the moment, I'm trying to get some fixes in before taking another stab at it (particularly since the problem with LocalTime not outputting a time zone string came up again the other day) - though the last time that I was just about ready to do it, I got stuck waiting for some stuff to get merged and then didn't have time.

**Comment by jmdavis** [comment]
Okay. I just updated the "warning" message to be clearer that it was specifically `LocalTime` and `UTC` which were affected, and I removed the part about when it was going away (just implying that that the behavior wasn't permanent so that no one would rely on it) on the theory that we're keeping that behavior around long term, as annoying as that is (though I added something to grep for later so that I know when to remove the warning from the docs, treating it like something that's being deprecated). No code was changed from what I had before.

**Comment by jmdavis** [comment]
Any chance of anyone looking at this and merging it soon? I know it's not a small change, but I was going to start in on splitting std.datetime after it was merged, and as long as this is here, I can't really do that. Though with my luck, I won't find the time to work on it anyway...

I should probably just stop trying to fix anything in std.datetime until I've actually found the time to split it and gotten it split and merged, since it seems like every time I might have the time to do it, I have a PR blocking it.

**Comment by schveiguy** [comment]
I think this looks fine. It's still hard to parse where all the code changes are. But I'd love to see datetime split up so it's easier to review.

I'm going to go based on the auto tester that this LGTM. If nobody comments in the next day, I'll merge.

**Comment by jmdavis** [comment]
Okay. Thanks. I can't promise anything, but I'm hoping to take a crack at splitting it this weekend, which is why I nagged about this PR.

I keep making smaller changes when I have some time available (but not enough to split std.datetime), and then those PRs block me when I do have a large enough block of time available that I might be able to do the split. So, I keep shooting myself in the foot. :|

**Comment by schveiguy** [comment]
Sorry, I forgot to pull this.

**Comment by schveiguy** [comment]
Auto-merge toggled on
